### DISCUSSION TITLE
Postgres Adapter... I do not get it.

### DISCUSSION THREAD

**Discussion by Inushin** [socketio/socket.io#4418] [Q&A]
Hi guys ^^
I have been swimming across the official SocketIO documentation but... I do not get it. I am working with two separated machines, both connected to another machine with a PostgreSQL DB. When I  run SocketIO on the same machine, everything is fine, it works as it has to. I add the adapter and the table creation that the documentation says is all ok. I am using sequelize sooo fine, the table is created and a  periodical DELETE and NOTIFY are launched to that table. Oooook. But that table, the place where magic should happen is... dry. There is no data input. NEVER. So the table is always empty and, as I can assume,  this is why when I launch the other machine, SocketIO from machine 1  does not link with SocketIO from machine2, each of the machines creates their separate rooms and are never related because the data that has to do it, is not being stored in the DB. Where I am messing up?

Apart from this, I think that it would be appreciated to add more documentation to that PostgreSQL adapter. Such as the one we have with the Redis Adapter. I suppose that the working behind is similar, but for newbies with SocketIO it could help us ^^

**Comment by darrachequesne** [comment]
Hi! The table might stay empty, if you don't emit packets whose size is above 8000 bytes or with binary data.

Please check https://socket.io/docs/v4/postgres-adapter/#how-it-works

**Reply by Inushin** [reply]
Yeah, but then, where is the usage of the adapter? I mean, how can I use it If it is correctly instantiated but "the way I think it works" with DB is only used in that specific case?

I mean...

Where I connect IO to the server, the adapter is launched and a table is created in the DB:
```
  
const startSocketIo = function (server) {
  try {
    const pool = new Pool({
      user: process.env.PGUSER,
      host: process.env.PGHOST,
      database: process.env.PGDATABASE,
      password: process.env.PGPASSWORD,
      port: process.env.PGPORT,
    });

    pool.query(`
      CREATE TABLE IF NOT EXISTS postgresSchema.socket_io_attachments (
          id          bigserial UNIQUE,
          created_at  timestamptz DEFAULT NOW(),
          payload     bytea
      );
    `);
  } catch (e) {
    console.error(e);
  }

  io = new Server(server);

  io.adapter(createAdapter(pool));
      io.of('/').adapter.on('create-room', (room) => {
        console.warn(`room ${room} was created`);
      });
      io.on('connection', async (socket) => {
        // Verify User JWT
        const userAuth = await jwt.verifyToken(socket.handshake.auth['token']);
        if (!userAuth) {
          return socket.disconnect();
        }
```
And the other place where the "adapter" itself is called. During the "streaming" of the text. "Clients" has the array with the rooms and socket IDs created in each client login.
```
const clients = await io.sockets.adapter.rooms;
```
What I do not understand is that... ok, you launch it in the same machine and all is OK (because that socket data is stored in the machine's cache memory, ok) but when you try to reach that data from another server by login in the same room... nothing, because the machine that is sending that data, is not saving it outside its machine. The adapter needs to be there for that. According to the documentation, the table it creates only saves data in some specific situations so... if that is not the use I suppose it has... why is it not doing the connection between those two machines?

**Reply by Inushin** [reply]
Apparently, in the local machine, that adapter creates the table `socket_io_attachments` normally, but in the dev server where I specify the sequelize schema, it creates that table in the "public" schema instead of the specified one. Can all this be related to a bad performance between the adapter and sequelize? The other queries that pass throw sequelize in other steps, work perfectly so this has just popped into my mind.

EDIT: You have to specify the schema with the table creation. (`schemaYouWant.socket_io_attachments`)

**Reply by rohitbrleaf** [reply]
I have users with uuid that joins the room , am taking all the users from postgresql . now i want to update the connected field to online whenerer a user joins and want to update connected to false when a user disconnects. am using this adapter but it fails to update in real time. do anyone have another solution for this?

**Reply by darrachequesne** [reply]
> but it fails to update in real time

@rohitbrleaf sorry to hear that. Could you reliably reproduce the issue? You can start with the example [there](https://github.com/socketio/socket.io-fiddle/tree/postgres-adapter).
